File: 프로그래머스/lv2/118667. 두 큐 합 같게 만들기/두 큐 합 같게 만들기.cpp
#include <string>
#include <vector>

using namespace std;
long tot, a, b, c, d, cnt;
int solution(vector<int> queue1, vector<int> queue2) {
    vector<int> v = queue1;
    for(int i = 0 ; i < queue1.size() ; i++){
        v.push_back(queue2[i]);
        a += queue1[i];
        b += queue2[i];
    }

    if((a + b) % 2 == 1)
        return -1;
    
    int i = 0, j = queue1.size();
    while (a != b){
        if(a > b){
            a -= v[i];
            b += v[i++];
        } else if(a < b) {
            b -= v[j];
            a += v[j++];
        } else
            break;
        if(j == queue1.size() * 2)
            j = 0;
        
        cnt++;
        if(cnt > queue1.size() * 2 + 2){
            cnt = -1;
            break;
        }
    }   
    

    int answer = cnt;
    return answer;
}

File: 프로그래머스/lv2/118667. 두 큐 합 같게 만들기/두 큐 합 같게 만들기.java
import java.util.*;

class Solution {
    public int solution(int[] queue1, int[] queue2) {
        Queue<Integer> q1 = new LinkedList<>();
        Queue<Integer> q2 = new LinkedList<>();
        Long total1 = 0L;
        Long total2 = 0L;
        
        
        for(int i = 0; i < queue1.length; i++) {
            q1.add(queue1[i]);
            total1 += queue1[i];
            q2.add(queue2[i]);
            total2 += queue2[i];
        }

        
        if((total1 + total2) % 2 == 1)
           return -1;
        if(total1 == total2)
            return 0;
        return solve(q1, q2, total1, total2);
    }
    
    
    public int solve(Queue<Integer> q1, Queue<Integer> q2, Long t1, Long t2) {
        int maxLength = q1.size() * 3;
        int cnt = 0;
        int idx = 0;
        Long target = (t1 + t2) / 2;
        
        
        while(idx < maxLength) {
            if(t1 > t2){
                int element = q1.poll();
                t1 -= element;
                t2 += element;
                q2.add(element);
                cnt++;
            }   else if(t1 < t2) {
                int element = q2.poll();
                t2 -= element;
                t1 += element;
                q1.add(element);
                cnt++;
            }
            else return cnt;
            idx++;            
        }
        
        
        return -1;
    }
}